src/ufunc.rs

```diff
@@ -0,0 +1,246 @@
+/// use ndarray::{azip, ArrayView1, ArrayViewMut1};
+/// use numpy::ufunc::{from_func, Identity};
+///
+/// Python::with_gil(|py| {
+///    let logit = |[p]: [ArrayView1<'_, f64>; 1], [α]: [ArrayViewMut1<'_, f64>; 1]| {
+///        azip!((p in p, α in α) {
+///            let mut tmp = *p;
+///            tmp /= 1.0 - tmp;
+///            *α = tmp.ln();
+///        });
+///    };
+///
+///    let logit =
+///        from_func(py, CString::new("logit").unwrap(), Identity::None, logit).unwrap();
+///
+///    py_run!(py, logit, "assert logit(0.5) == 0.0");
+///
+///    let np = py.import("numpy").unwrap();
+///
+///    py_run!(py, logit np, "assert (logit(np.full(100, 0.5)) == np.zeros(100)).all()");
+/// });
+/// ```
+///
+/// ```
+/// # #![allow(mixed_script_confusables)]
+/// # use std::ffi::CString;
+/// #
+/// use pyo3::{py_run, Python};
+/// use ndarray::{azip, ArrayView1, ArrayViewMut1};
+/// use numpy::ufunc::{from_func, Identity};
+///
+/// Python::with_gil(|py| {
+///     let cart_to_polar = |[x, y]: [ArrayView1<'_, f64>; 2], [r, φ]: [ArrayViewMut1<'_, f64>; 2]| {
+///         azip!((&x in x, &y in y, r in r, φ in φ) {
+///             *r = f64::hypot(x, y);
+///             *φ = f64::atan2(x, y);
+///         });
+///     };
+///
+///     let cart_to_polar = from_func(
+///         py,
+///         CString::new("cart_to_polar").unwrap(),
+///         Identity::None,
+///         cart_to_polar,
+///     )
+///     .unwrap();
+///
+///     let np = py.import("numpy").unwrap();
+///
+///     py_run!(py, cart_to_polar np, "np.testing.assert_array_almost_equal(cart_to_polar(3.0, 4.0), (5.0, 0.643501))");
+///
+///     py_run!(py, cart_to_polar np, "np.testing.assert_array_almost_equal(cart_to_polar(np.full((10, 10), 3.0), np.full((10, 10), 4.0))[0], np.full((10, 10), 5.0))");
+///     py_run!(py, cart_to_polar np, "np.testing.assert_array_almost_equal(cart_to_polar(np.full((10, 10), 3.0), np.full((10, 10), 4.0))[1], np.full((10, 10), 0.643501))");
+/// });
+/// ```
+pub fn from_func<'py, T, F, const NIN: usize, const NOUT: usize>(
+    py: Python<'py>,
+    name: CString,
+    identity: Identity,
+    func: F,
+) -> PyResult<&'py PyAny>
+where
+    T: Element,
+    F: Fn([ArrayView1<'_, T>; NIN], [ArrayViewMut1<'_, T>; NOUT]) + 'static,
+{
+    let wrap_func = [Some(wrap_func::<T, F, NIN, NOUT> as _)];
+
+    let r#type = T::get_npy_type().expect("universal function only work for built-in types");
+
+    let inputs = [r#type as _; NIN];
+    let outputs = [r#type as _; NOUT];
+
+    let data = Data {
+        func,
+        wrap_func,
+        name,
+        inputs,
+        outputs,
+    };
+
+    let data = Box::leak(Box::new(data));
+
+    unsafe {
+        py.from_owned_ptr_or_err(PY_UFUNC_API.PyUFunc_FromFuncAndData(
+            py,
+            data.wrap_func.as_mut_ptr(),
+            data as *mut Data<F, NIN, NOUT> as *mut c_void as *mut *mut c_void,
+            data.inputs.as_mut_ptr(),
+            /* ntypes = */ 1,
```

Comment:
Supporting the case where `ntypes > 1` will be a good future improvement (maybe not very difficult given what this PR already implemented, but may require lots of boilerplates)